### PR TITLE
Add Rollup ILM Action

### LINKED ISSUE #48003: Add an ILM action to rollup an index
Rollup should no longer be a continuous running job. Instead, it should be an action that can be triggered on ILM-managed indices. The action should iterate over all grouping tuples (dimensions) and calculate aggregate metrics (`min`/`max`/`avg`/`value_count`) generating one document per grouping tuple (dimension).

In the context of this issue we should implement the following:
- Implement a "grouping tuple" field type to generate the tuple at index time for the document.
- Implement a "rollup metric" field type to store aggregation results (min/max/count/avg etc). This metric should provide the correct information to requesting aggregator (e.g. `avg` agg on `rollup_metric` will internally fetch sum + count)
- Store document count somewhere on the document (docvalue field?)
- Create an ILM rollup action

### PR DESCRIPTION
**Pull request by talevy**
this commit introduces a new Rollup ILM Action that allows indices
to be rolled up according to a specific rollup config. The
action also allows for the new rolled up index to be associated with
a different policy than the original/source index.

Relates #42720.

Closes #48003.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (not-napoleon) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:75]: I realize this is a matter of taste, but this function name really sounds like it's intended to take an action (delete), not return a flag value.  Normally I don't care, but in this case I really think `getDeleteOriginalIndex()` (or maybe `shouldDeleteOriginalIndex`) would be a better name.

**Review by not-napoleon — APPROVED** [review]
This looks good to me, but probably worth having someone from ILM review as well.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:75]: totally valid! I think I sometimes find myself going away from what I prefer: having boolean instance-get method names to be `is<varName>`. I will rename 👍

**Comment by talevy** [comment]
run elasticsearch-ci/default-distro

**Review by csoulios — COMMENTED** [review]
This is great work and it looks a lot simpler than I had expected. 
I left some comments wrt to rollup specific points for discussion.
Also, I am no expert in ILM. So, perhaps someone with more experience in this area could help

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:65]: yeah... I consider that something we can work on in a follow-up. it will require some more conditionals etc that I thought would distract this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:117]: I will double-check with the ILM team, but this is what most of the other actions are doing

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java:44]: I am not sure this needs to be configurable? Specifically if this is used against a data-stream, then this would not be a concern. We can discuss! I would like to reduce the concerns, but if this type of conflict is expected, we do need a work-around for users. you're right

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:35]: sure, will rename!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java
Comment (csoulios) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:65]: It makes sense. Thanks

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (csoulios) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:34]: After our last discussion on defining the ILM policy of the rollup index separately. Is this still the place to set the rollup index poly?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java
Comment (csoulios) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:45]: should `RollupILMAction` be added in a `if (RolupV2.isEnabled() {.....}` block?

Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:45]: yes, indeed! I slipped here because I was eager to see CI run the integration tests. will fix

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java:44]: We use prefixes for most of the other ILM actions (searchable snapshot and shrinking), I think it'd be nice to stick with that convention for rollup indices also

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java:48]: What happens if `response.isAcknowledge()` is false here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java:57]: I think these can be split into two steps, that way the update settings request can be retryable, rather than a failure of that step causing the entire rollup to fail

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java:38]: I think it would be really nice to have this be retryable for the first release, maybe we can brainstorm why it isn't retryable and how to make it so? Otherwise I think we may end up having to do surgery on the indices if they get into an error and can't proceed past the rollup stage

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java:105]: This should use the `ReadOnlyStep` (which uses the new API for marking things as read only) rather than updating the settings manually

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some general comments for this. I would like to get it retryable and maybe go through the different failure scenarios for how this could be cleaned up to make it retryable. Maybe someone can elaborate on how it can fail and how it will need to be cleaned up?

Another question I have is about the rollup policy, do we currently write custom metadata to an index that marks it as a "rolled up" index? If so, I think we may want to inspect that metadata in other places within ILM (rollover is the big one that comes to mind) so that we can skip steps that are not needed in the event that a user specifies a custom policy for the new index where the policy contains a rollover action (since rollup indices are read only they would have no need for rollover).

**Comment by talevy** [comment]
> Another question I have is about the rollup policy, do we currently write custom metadata to an index that marks it as a "rolled up" index? If so, I think we may want to inspect that metadata in other places within ILM (rollover is the big one that comes to mind) so that we can skip steps that are not needed in the event that a user specifies a custom policy for the new index where the policy contains a rollover action (since rollup indices are read only they would have no need for rollover).

we do have custom metadata for rolled-up indices, but intend on only making this important for search-resolution, and nothing else should depend on it.

but you bring up a good point, I think adding a step to check rollup makes sense, since some actions do not make sense -- like rolling-over a rolled-up index.

thanks for the feedback, I'll work on those. Some of the details of this metadata are up-in-the-air at the moment, but a usable-version exists today

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupStep.java]: there is more work to do with regards to the RollupAction itself, which will evolve the response from the action. For now it either throws exception or returns `true` as a placeholder before we introduce more subtle edge-case behaviors (e.g. #66423)

**Comment by talevy** [comment]
run elasticsearch-ci/2

**Review by dakrone — CHANGES_REQUESTED** [review]
I left two more really small changes. We need to add the rollup step to the `IGNORE_STEPS_MAINTENANCE_REQUESTED` steps to prevent the node from being terminated during a rollup, and we also need to add documentation for the new rollup step.

File: docs/reference/ilm/ilm-actions.asciidoc
Comment (talevy) [inline docs/reference/ilm/ilm-actions.asciidoc:73]: Hey @jrodewig, mind letting me know if I did this right? thanks!

Comment (jrodewig) [reply docs/reference/ilm/ilm-actions.asciidoc:73]: This looks correct to me!

File: docs/reference/ilm/actions/ilm-rollup.asciidoc
Comment (jrodewig) [inline docs/reference/ilm/actions/ilm-rollup.asciidoc:10]: I don't think we need the caps here.
```suggestion
For more information about rollup, see the <<rollup-api, rollup action documentation>>
```

File: docs/reference/ilm/actions/ilm-rollup.asciidoc
Comment (jrodewig) [inline docs/reference/ilm/actions/ilm-rollup.asciidoc:20]:
```suggestion
The rollup configuration, a more detailed description of the
rollup configuration specification can be found <<rollup-api-request-body,here>>.
```

**Review by jrodewig — APPROVED** [review]
Docs LGTM. Thanks @talevy!

File: docs/reference/ilm/actions/ilm-rollup.asciidoc
Comment (jrodewig) [inline docs/reference/ilm/actions/ilm-rollup.asciidoc:24]: Removes a duplicate space.
```suggestion
The name of an <<index-lifecycle-management, {ilm}>> ({ilm-init}) policy to associate
```

File: docs/reference/ilm/ilm-actions.asciidoc
Comment (talevy) [reply docs/reference/ilm/ilm-actions.asciidoc:73]: thanks!

**Comment by talevy** [comment]
thanks all! ready for another review!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:64]: I missed this before, but if we're going to rollup right after a shrink and force merge, what's the point of shrinking and force merging? Should we just rollup first and then do the shrink/forcemerge?

(I think this is okay to leave as is for now, but we might want to investigate making them mutually exclusive in the future so we don't waste resources doing a shrink and force merge right before we rollup and create a new index anyway. At the very least I think maybe force merge should move to after the Rollup?)

**Review by dakrone — APPROVED** [review]
This LGTM, thanks for iterating on this Tal!

I left one more comment, but it can be addressed either here or in a subsequent PR (if we decide a change is desired), up to you.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java
Comment (talevy) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java:64]: these are fair points. I figured these things would occur anyways, so order should not matter at the end of the day. That being said, waiting to do these operations before rolling up is not necessary. it might be interesting to benchmark and see if fewer shards ends up resulting in a faster rollup or not. but this should not be of concern to the user anyways.

I will move Rollup to happen before Shrink, but will bring this up with the team to see if there are any other opinions

File: docs/reference/ilm/actions/ilm-rollup.asciidoc
Comment (ruflin) [inline docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: @talevy How does this work in the context of data streams? Does it work in the context of data streams? What are the index names there?

Comment (jrodewig) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: ILM works on the index level, not directly on data streams. If the index is a backing index for a data stream, the rollup index is a backing index for the same stream. The name of the new rollup index is based on the name of the backing index. For example, `.ds-my-data-stream-2021.01.29-000001` would become `rollup-.ds-my-data-stream-2021.01.29-000001`.

This is similar to how the ILM `shrink` action currently renames backing indices:

https://www.elastic.co/guide/en/elasticsearch/reference/master/data-streams.html#data-streams-generation

I hope that helps. I'll open up a follow-up PR to better clarify this in the docs.

Comment (talevy) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: Hey @ruflin and @jrodewig,

this is exactly how it works today, but due to some failure-scenarios we discussed recently, we're going to change this behavior.

In the context of datastreams, we will likely randomize the name of the rollup index to prevent collision and conflict with any past or rogue rollup actions. The belief is that the underlying name of the index should not be of concern to users since they will be added into the datastream and managed in ILM. Do you see any problems with this on your end @ruflin?

For this reason, it might make sense to hold off just a little bit on your docs change PR @jrodewig, but much appreciated!

I've opened #68225 to track it. Will likely change next week

Comment (jrodewig) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: Thanks for the update, @talevy!

Comment (ruflin) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: > The belief is that the underlying name of the index should not be of concern to users since they will be added into the datastream and managed in ILM

If this is the case, it is great. I was under the assumption all indices belong to a data stream must use the `.ds-*` prefix and be hidden. So I expect index names like `.ds-rollup-*` or something similar. This would also ensure there are no conflicts and makes it clear that these rollups belong to the data streams.

The `rollup-` prefix is what put me on the wrong track here and made me assume it does not belong to the data stream.

Comment (talevy) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: it is actually common for indices managed by ILM to be renamed with a new prefix. For example, shrunken indices start with the `shrunk-` prefix. 

independent of the name, the indices will be hidden and added to the backing datastream

Comment (ruflin) [reply docs/reference/ilm/actions/ilm-rollup.asciidoc:12]: @talevy Interesting. Why are these not prefixed by a `.` and not inside `.ds-*-shrunk-*` or similar to indicate they belong together? I think I'm derailing but could you share a link where I can read up on this? The part I'm concerned is that this not seem to fully fit into our naming scheme: https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme

**Pull request merged**
